File: backend/app/schemas/settings_github.py
"""GitHub 集成设置 API 契约。"""

from pydantic import BaseModel, Field


class GithubSettingsRead(BaseModel):
    """GET 响应：不返回完整 Token。"""

    has_token: bool = Field(description="是否存在可用 Token（数据库或环境变量）")
    token_preview: str | None = Field(
        default=None,
        description="仅当 Token 保存在数据库时返回末 4 位；仅从环境变量读取时不返回末位",
    )
    token_length: int | None = Field(
        default=None,
        description="仅当 Token 保存在数据库时返回字符长度；仅从环境变量读取时不返回",
    )


class GithubSettingsUpdate(BaseModel):
    token: str | None = Field(
        default=None,
        description="写入数据库；传 null 或空字符串表示清除数据库中的 Token",
    )


class GithubTestRequest(BaseModel):
    token: str | None = Field(
        default=None,
        description="可选：测试指定 Token；省略则测试当前生效 Token（数据库或环境变量）",
    )


class GithubProfileRead(BaseModel):
    login: str = Field(description="GitHub 用户名")
    name: str | None = Field(default=None, description="显示名称")
    avatar_url: str = Field(description="头像 URL")
    html_url: str = Field(description="GitHub 主页")


class GithubTestResponse(BaseModel):
    ok: bool
    message: str | None = None

File: frontend/src/lib/settings-github.ts
export type GithubSettingsRead = {
  has_token: boolean
  token_preview: string | null
  token_length: number | null
}

export type GithubSettingsUpdate = {
  token: string | null
}

export type GithubProfileRead = {
  login: string
  name: string | null
  avatar_url: string
  html_url: string
}

export type GithubTestRequest = {
  token?: string | null
}

export type GithubTestResponse = {
  ok: boolean
  message?: string | null
}

async function parseErrorMessage(res: Response): Promise<string> {
  try {
    const data: unknown = await res.json()
    if (data && typeof data === "object" && "detail" in data) {
      const d = (data as { detail: unknown }).detail
      if (typeof d === "string") {
        return d
      }
      return JSON.stringify(d)
    }
  } catch {
    // ignore
  }
  return res.statusText || `HTTP ${res.status}`
}

export async function fetchGithubSettings(): Promise<GithubSettingsRead> {
  const res = await fetch("/api/settings/github")
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<GithubSettingsRead>
}

export async function putGithubSettings(body: GithubSettingsUpdate): Promise<GithubSettingsRead> {
  const res = await fetch("/api/settings/github", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<GithubSettingsRead>
}

export async function fetchGithubProfile(): Promise<GithubProfileRead> {
  const res = await fetch("/api/settings/github/profile")
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<GithubProfileRead>
}

export async function postGithubTest(token?: string): Promise<GithubTestResponse> {
  const body: GithubTestRequest | undefined =
    token !== undefined && token.trim() !== "" ? { token: token.trim() } : undefined
  const res = await fetch("/api/settings/github/test", {
    method: "POST",
    headers: body ? { "Content-Type": "application/json" } : undefined,
    body: body ? JSON.stringify(body) : undefined,
  })
  if (!res.ok) {
    throw new Error(await parseErrorMessage(res))
  }
  return res.json() as Promise<GithubTestResponse>
}
